### ISSUE TITLE
[no-misused-promises] false positives with overloaded function

### ISSUE THREAD

**Issue by AviVahl** [typescript-eslint/typescript-eslint#4609]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

- https://github.com/AviVahl/typescript-eslint-bug
- default tsconfig from `tsc --init`
- minimal eslint config
- clone and `npm it` to see issue

**Expected Result**

mocha's `it` accepts a return promise in its callback type, so there should be no error.

**Actual Result**

```
typescript-eslint-bug/test/some-test.spec.ts
  2:32  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```
mishandling of overloads defined on an interface?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.13.0` |
| `@typescript-eslint/parser`        | `5.13.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.14.0` |

**Comment by AviVahl** [comment]
Just made reproduction repository public (my bad).

**Comment by JoshuaKGoldberg** [comment]
Saving some digging time for future us:

```
npm i @types/mocha --save-dev
```

```ts
describe("suite", () => {
  it("checks something async", async () => {
    //
  });
});
```

**Comment by bradzacher** [comment]
Okay soooo in mocha the `it` function is defined using this type:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/631dd9c407e21a97b944e6e1fa9cafb0b854d177/types/mocha/index.d.ts#L490-L505

[It supports BOTH `=> void` and `=> PromiseLike<any>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/631dd9c407e21a97b944e6e1fa9cafb0b854d177/types/mocha/index.d.ts#L2281-L2289) as its callback.

So it looks like this might be a bug in that we don't properly consider overloaded signatures?

**Comment by bradzacher** [comment]
Confirming my previous comment, [here is a playground which shows this incorrect behaviour of not respecting overloads](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsAgPYkC2cUArlBACYlb7UOtQBcYAOQR83fELAhgAKDlsIuGAEN8EMLmoIoKAGYIAjIIAUuAEYmAlGAC8APjAA3anDbX7Tl2wDcM4MEgADxxcJHZIMWp8GX0DY2N3BwAFbl4IADo1KGoYRwgEy18ZBSVVdU1tPQQAJhNzK1tk1MYIAB5nVztEz1dffyCQsLYwGgjxGJr47pSeFszoHLyCorlEMPxdZVx1AHk8-BhqZTYAMUMwAG8ZMBuwUws77o63QWffW7v6x8awGbT2rxdV5eXwAX3kihUag0Wh0YGo+0OxzORjAe1ESNOhj6ASYCAgwUUQzGURkCIxRyxcQSPz+cyyi3ylkKclZbPZHM5rP8PN+ACUdgBZACSAGUAKJgAAi4pOorAJx2fN+ABkAIIATQA4gKAKoAOSlYAAKjswAB1JUAaTkBKwURQF1BRRKUPUmBwv2arAAwrCkPgmKEoiq4ABrdQ2EYQADuYFaxrsxgJiiYSCiJgZuQgJkcyhgTBzJrAAB8vbNWKGIwmuj9ngAaMBqABWRMzEGU2QQAH5BMoEBgnl4h65pt6IFW2omimtRJttuW0pOa5drrdQCA17dwGqkEgtgALaAafMwMxbMNQMC6KJgJBHpsLAtIOBaMD9tjAW8tomvhDw3Q7wfOlWHSLcbnAVAsFUZQ6HhBBdCYGBdDgGAYHCY0H1wU9z1wMM72oIJUzCMAYyPf973UED1EYR9smzNgwI+KQ0Gg-BYPgn9QgwrCcIvAiiNwNN1DIiAKOA8cwForihiYj5ILUJAmHwbQwDVRcWmvW9KJhOgsHQl832oQCyLgXAD0gfYTzQ3D8NolMhNk8CpFkD5KIQBM+WgJCkAMWwTUbY0vOYGAkGqfz8X2JMtEQ5DUPQthezuYw8wLItjW6ILvNCvyy2o5cspC3zazLPEFBQ-FhjLBAkJgRstBk9gkviNROy0PsB0y4KfPCvLxwK7rQuqEqwDKiAKvCarassQR8vDKdBt80sTUW4awTkDdrnALyuGgMSkCvHTND0gy-wA99-07DAEHM7gaBYeEcHYwyEC22RZw2LYqPHFcrg+TbmJ3PdD2PbCbIvK8b3wID1CzNMzo-L9oZks7jJhjTQOcyC2I4mKkJQtCePKPi8IEhzhNI8j0eoqSryzPJGKx1iYLghqIFbbjhkw4nwdJ9NBIp0TxO+isaLp9miXYOTtzQRTlNU9Saah9Hjv0iAXvO0zzMs0RrLPfj7MCYipax1zbnczzsqWqNjUC1aIogKLjDxuLCcSkwUvzQtBAyn5Cp83KMYnebLaKgwRrGiaqpGWr6oQRr3eS1quw6wc-ftvrRYGq3husUqEHKxBJpjtCZqD0OA+W-2hrsIp-pATdAbU4HzOPZQ9dsrToa0GAMHRlGjMAnTqOliDmfY1n44lzmTV43n8P58mSKF6nJOk6enKb1A5ZUq9Fck5Wjp4NWNbRo6SbDUeXPA7CkHMivQodp22Y5oZmuMZP2outOPGrlBM6XCHP+EcC7jSLtHGqpdZo-WNFXRatcZDghkB9ecIs0h+kqIGYM0M-rrgbkzdSahdCiDEgufmOkuDUHTB6DIptwKtTYD3PulDqHYCLNRVo-YMAIPAgDeSYAfStTCFedu+I4wjyZlBFmAt0zQ3UmDfWpMWDhH5ogOAL58xwAAF7ugfJQtI6RZ60QUZ3Wi0EoCsGGEgGMhFVAAHMmB0H2gIJm7d6Y80UfhZRVjCLuPRvozSMZ1EWXbqlQs8JoY6SyD5c6-YqFHmhgE1g9ZXEF3fI+V+Hd+LeIEjJfxkkgn3nSZQxwrhwif3-LeUQ4gr7SHAmI+MiZkxG0chmO47jcxe3SstOa1ZEwjjYI2GS7Y2o9lTgMiZ0DRY1jrngxu-DBEdmEekmm95lAoHXvRBmpFgkXXfGIZQfc0bRNCleFep5zoUO4KUhQwxqJQCZu4+qyMN7hHOQOIOtMMmS0ZlvHGcEwnHjVJdA5RzAL3NqbLdW8s97RnEeOWpZsbingTEEMIBdxbHCYaNBAYYaAxgQAAbQALrLRJUmQFAgTRlw4RcMAJAGHYsJUkKS-4IwYDRsaYlgg1QxmUOo9gCZmXEocKCHh-0AisCLPAMw6RoDVAAAwGAAKzpAFexMw6F0iMQOrwgIKKmmUsEMKdYyhNVTh6f1IBdggTl2NOS2ZNw+Ey0Westuny1kbPFls8I35XnDHeWC-JCIyl3PHCI6E7i2BMz9a-E2fzpGUrUiC9iQaIWSJ3grOFQdEX0K+qigk6K2CYsYQgXuOK8XUAJSSslIrPZpSpRlKZ-xeX8qGEKmqdAzCiBFeK9ckqIDSrgLK+VSrVXqrNVqnVDyJVNnzYarpVKTWiEnVOW1HDYGFrEsWoOy5EDEOhrqBw3YwC6jAD7PtTr8FN1dcs0RsZvkzySeoQ+ejrmhsfKMyFUiJ6fq7LPWGHZ-1XNFqRTsj7N78O3tC3eals0JxzXQj4MlUVRkiqIJMFSkpcJpTAy9YBnVj1ve6hpUawDPu-Zm2FZGKNIduO4hIzaWgAk6I6gj16FlCJIw+mjB9tJvpDbcpwXTv3-OE2lODtGt5UckwUg81BWCsvWAgfMYAdBurAHQdZrdDoCZueEKT8kkV0UZDWet3tiwAJaAVddeH1rFEhGUYT0NqJMd9P6LBsjvBAA&rules=N4XyA&tsConfig=N4XyA)

**Comment by dietergeerts** [comment]
I just upgraded to ` 5.13.0` hoping this issue was fixed, but it isn't. I still get the errors on mocha test functions.

![image](https://user-images.githubusercontent.com/7393706/156755606-2f0d0694-7d78-4480-ab02-1f4450deeff2.png)
![image](https://user-images.githubusercontent.com/7393706/156755628-05441bb0-d43d-473d-8499-a0e85d2c7fe4.png)

From the changelogs, I see that this should be in the `5.13.0` version, or am I missing something?
